Implementation of statistical gas wall boundary conditions

toquydong opened the pull request:
**Summary**

In these fixes, some stochastic boundary conditions which interest the gas simulation community, are implemented in LAMMPS. These type of BCs originated from the kinetic theory (e.g BCs of Boltzmann equations) are very useful for simulating gas transport in a solid pore. Instead of using explicitly the atomic wall, it is sufficient to model statistically the independent collisions between the wall and the gas molecules. This will save a lot of computation cost devoted to the wall. During the simulation, if a collision is detected, e.g a gas atom cross the wall plane, it is reinjected into the interior domain with random velocity following special distributions.

Three stochastic models are : Diffusive wall model, Maxwell model,  Cercignani Lampis model. Those are popular parametric models based on accommodation parameters.


**Related Issues**

No issues.

**Author(s)**

Quy-Dong To, Université Paris Est Marne la Vallée. Email: toquydong@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

It is compatible backwards.

**Implementation Notes**

The wall reflection velocity is random. There are statistical models for this effect.

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [x ] The feature or features in this pull request is complete
- [ x] Licensing information is complete
- [x ] Corresponding author information is complete
- [ x] The source code follows the LAMMPS formatting guidelines
- [ x] Suitable new documentation files and/or updates to the existing docs are included
- [ x] The added/updated documentation is integrated and tested with the documentation build system
- [ x] The feature has been verified to work with the conventional build system
- [ x] The feature has been verified to work with the CMake based build system

**Further Information, Files, and Links**

No further information.

akohlmey commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+
+  if (reflectionstyle != NONE) {
+
+    do seedfix = rand(); while (seedfix > 900000000);
```
this piece of code is not a good idea. elsewhere in LAMMPS, there is a random seed is required as input which allows for calculations to be reproducible for debugging purposes.

toquydong commented:
Hi Axel,
Thanks for the corrections you made to the code and the doc file.
Regarding your comments, should I introduce a random seed input in the 
fix command?
Or several random seeds  depending on the number of the stochastic wall 
present in the model?

Best regards,
Quy Dong
Le 24/09/2019 20:32, Axel Kohlmeyer a écrit :
>
> *@akohlmey* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/USER-MISC/fix_wall_stochastic.cpp 
> <https://github.com/lammps/lammps/pull/1691#discussion_r327770698>:
>
> > +
> +  //RanMars *random;
> +  if (strcmp(arg[3],"diffusive") == 0) {
> +    reflectionstyle = DIFFUSIVE;
> +    arginc = 6;
> +  } else if (strcmp(arg[3],"maxwell") == 0) {
> +    reflectionstyle = MAXWELL;
> +    arginc = 7;
> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
> +    reflectionstyle = CERCIGNANILAMPIS;
> +    arginc = 9;
> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
> +
> +  if (reflectionstyle != NONE) {
> +
> +    do seedfix = rand(); while (seedfix > 900000000);
>
> this piece of code is not a good idea. elsewhere in LAMMPS, there is a 
> random seed is required as input which allows for calculations to be 
> reproducible for debugging purposes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lammps/lammps/pull/1691?email_source=notifications&email_token=ANFYXKOQKYY674NRF7A3Y6TQLJMK3A5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFYRT3Q#pullrequestreview-292624878>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ANFYXKLFSSNJHFEEWKH7UHTQLJMK3ANCNFSM4IZNCFFQ>.
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

toquydong commented:
For the latter case, there will be a lot of input parameters in the fix 
input.


Le 25/09/2019 09:58, TO Quy Dong a écrit :
> Hi Axel,
> Thanks for the corrections you made to the code and the doc file.
> Regarding your comments, should I introduce a random seed input in the 
> fix command?
> Or several random seeds  depending on the number of the stochastic 
> wall present in the model?
>
> Best regards,
> Quy Dong
> Le 24/09/2019 20:32, Axel Kohlmeyer a écrit :
>>
>> *@akohlmey* commented on this pull request.
>>
>> ------------------------------------------------------------------------
>>
>> In src/USER-MISC/fix_wall_stochastic.cpp 
>> <https://github.com/lammps/lammps/pull/1691#discussion_r327770698>:
>>
>> > +
>> +  //RanMars *random;
>> +  if (strcmp(arg[3],"diffusive") == 0) {
>> +    reflectionstyle = DIFFUSIVE;
>> +    arginc = 6;
>> +  } else if (strcmp(arg[3],"maxwell") == 0) {
>> +    reflectionstyle = MAXWELL;
>> +    arginc = 7;
>> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
>> +    reflectionstyle = CERCIGNANILAMPIS;
>> +    arginc = 9;
>> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
>> +
>> +  if (reflectionstyle != NONE) {
>> +
>> +    do seedfix = rand(); while (seedfix > 900000000);
>>
>> this piece of code is not a good idea. elsewhere in LAMMPS, there is 
>> a random seed is required as input which allows for calculations to 
>> be reproducible for debugging purposes.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/lammps/lammps/pull/1691?email_source=notifications&email_token=ANFYXKOQKYY674NRF7A3Y6TQLJMK3A5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFYRT3Q#pullrequestreview-292624878>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe-auth/ANFYXKLFSSNJHFEEWKH7UHTQLJMK3ANCNFSM4IZNCFFQ>.
>>
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

akohlmey commented:
> Hi Axel, Thanks for the corrections you made to the code and the doc file. Regarding your comments, should I introduce a random seed input in the fix command? Or several random seeds depending on the number of the stochastic wall present in the model? Best regards, Quy Dong Le 24/09/2019 20:32, Axel Kohlmeyer a écrit :
> […](#)
> ***@***.**** commented on this pull request. ------------------------------------------------------------------------ In src/USER-MISC/fix_wall_stochastic.cpp <[#1691 (comment)](https://github.com/lammps/lammps/pull/1691#discussion_r327770698)>: > + + //RanMars *random; + if (strcmp(arg[3],"diffusive") == 0) { + reflectionstyle = DIFFUSIVE; + arginc = 6; + } else if (strcmp(arg[3],"maxwell") == 0) { + reflectionstyle = MAXWELL; + arginc = 7; + } else if (strcmp(arg[3],"cercignanilampis") == 0) { + reflectionstyle = CERCIGNANILAMPIS; + arginc = 9; + } else error->all(FLERR,"Illegal fix wall/stochastic command"); + + if (reflectionstyle != NONE) { + + do seedfix = rand(); while (seedfix > 900000000); this piece of code is not a good idea. elsewhere in LAMMPS, there is a random seed is required as input which allows for calculations to be reproducible for debugging purposes. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#1691?email_source=notifications&email_token=ANFYXKOQKYY674NRF7A3Y6TQLJMK3A5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFYRT3Q#pullrequestreview-292624878>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ANFYXKLFSSNJHFEEWKH7UHTQLJMK3ANCNFSM4IZNCFFQ>.
> --- L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast. https://www.avast.com/antivirus

just one seed for all is sufficient. we only need to be able to reproduce a trajectory with the same number of processors and the same system with the same walls.

toquydong commented:
OK. I have made changes in the stochasticwall branch. I think it is ready.

Best regards,
Quy Dong
Le 25/09/2019 19:35, Axel Kohlmeyer a écrit :
>
>     Hi Axel, Thanks for the corrections you made to the code and the
>     doc file. Regarding your comments, should I introduce a random
>     seed input in the fix command? Or several random seeds depending
>     on the number of the stochastic wall present in the model? Best
>     regards, Quy Dong Le 24/09/2019 20:32, Axel Kohlmeyer a écrit :
>     … <#>
>     /*@*/.**** commented on this pull request.
>     ------------------------------------------------------------------------
>     In src/USER-MISC/fix_wall_stochastic.cpp <#1691 (comment)
>     <https://github.com/lammps/lammps/pull/1691#discussion_r327770698>>:
>     > + + //RanMars *random; + if (strcmp(arg[3],"diffusive") == 0) {
>     + reflectionstyle = DIFFUSIVE; + arginc = 6; + } else if
>     (strcmp(arg[3],"maxwell") == 0) { + reflectionstyle = MAXWELL; +
>     arginc = 7; + } else if (strcmp(arg[3],"cercignanilampis") == 0) {
>     + reflectionstyle = CERCIGNANILAMPIS; + arginc = 9; + } else
>     error->all(FLERR,"Illegal fix wall/stochastic command"); + + if
>     (reflectionstyle != NONE) { + + do seedfix = rand(); while
>     (seedfix > 900000000); this piece of code is not a good idea.
>     elsewhere in LAMMPS, there is a random seed is required as input
>     which allows for calculations to be reproducible for debugging
>     purposes. — You are receiving this because you authored the
>     thread. Reply to this email directly, view it on GitHub <#1691
>     <https://github.com/lammps/lammps/pull/1691>?email_source=notifications&email_token=ANFYXKOQKYY674NRF7A3Y6TQLJMK3A5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFYRT3Q#pullrequestreview-292624878>,
>     or mute the thread
>     https://github.com/notifications/unsubscribe-auth/ANFYXKLFSSNJHFEEWKH7UHTQLJMK3ANCNFSM4IZNCFFQ.
>     --- L'absence de virus dans ce courrier électronique a été
>     vérifiée par le logiciel antivirus Avast.
>     https://www.avast.com/antivirus
>
> just one seed for all is sufficient. we only need to be able to 
> reproduce a trajectory with the same number of processors and the same 
> system with the same walls.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lammps/lammps/pull/1691?email_source=notifications&email_token=ANFYXKPTHHYSMRYX4QJI5GDQLOOOZA5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7SXBFQ#issuecomment-535130262>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ANFYXKIF3N7PYHJ344PMOXTQLOOOZANCNFSM4IZNCFFQ>.
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

akohlmey commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+  } else error->all(FLERR,"Illegal fix wall/stochastic command");
+
+  if (reflectionstyle != NONE) {
+    seedfix = force->numeric(FLERR,arg[4]);
```
Since seedfix is an integer, this should be calling `force->inumeric()` .

akohlmey commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+
+  if (reflectionstyle != NONE) {
+    seedfix = force->numeric(FLERR,arg[4]);
+    if (seedfix > 900000000) error->all(FLERR,"Random seed must be less than 900000000");
```
why may the seed not be larger than 900000000?

akohlmey replied:
instead there should be a test not allowing a seed of 0 or negative. See, e.g., fix langevin for comparison.

akohlmey commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+  if (reflectionstyle != NONE) {
+    seedfix = force->numeric(FLERR,arg[4]);
+    if (seedfix > 900000000) error->all(FLERR,"Random seed must be less than 900000000");
+    random = new RanMars(lmp,seedfix);
```
This should be using `random = new RanMars(lmp,seedfix + comm->me);`. Or else random numbers may be correlated between different MPI ranks. See fix langevin.

toquydong commented:
I have modified the cpp file.

Quy Dong
Le 25/09/2019 22:27, Axel Kohlmeyer a écrit :
>
> *@akohlmey* requested changes on this pull request.
>
> I think this needs some more tweaks. Please see my comments.
>
> ------------------------------------------------------------------------
>
> In src/USER-MISC/fix_wall_stochastic.cpp 
> <https://github.com/lammps/lammps/pull/1691#discussion_r328322624>:
>
> > +  reflectionstyle = NONE;
> +
> +
> +  if (strcmp(arg[3],"diffusive") == 0) {
> +    reflectionstyle = DIFFUSIVE;
> +    arginc = 6;
> +  } else if (strcmp(arg[3],"maxwell") == 0) {
> +    reflectionstyle = MAXWELL;
> +    arginc = 7;
> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
> +    reflectionstyle = CERCIGNANILAMPIS;
> +    arginc = 9;
> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
> +
> +  if (reflectionstyle != NONE) {
> +    seedfix = force->numeric(FLERR,arg[4]);
>
> Since seedfix is an integer, this should be calling |force->inumeric()| .
>
> ------------------------------------------------------------------------
>
> In src/USER-MISC/fix_wall_stochastic.cpp 
> <https://github.com/lammps/lammps/pull/1691#discussion_r328323002>:
>
> > +
> +
> +  if (strcmp(arg[3],"diffusive") == 0) {
> +    reflectionstyle = DIFFUSIVE;
> +    arginc = 6;
> +  } else if (strcmp(arg[3],"maxwell") == 0) {
> +    reflectionstyle = MAXWELL;
> +    arginc = 7;
> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
> +    reflectionstyle = CERCIGNANILAMPIS;
> +    arginc = 9;
> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
> +
> +  if (reflectionstyle != NONE) {
> +    seedfix = force->numeric(FLERR,arg[4]);
> +    if (seedfix > 900000000) error->all(FLERR,"Random seed must be less than 900000000");
>
> why may the seed not be larger than 900000000?
>
> ------------------------------------------------------------------------
>
> In src/USER-MISC/fix_wall_stochastic.cpp 
> <https://github.com/lammps/lammps/pull/1691#discussion_r328323672>:
>
> > +
> +  if (strcmp(arg[3],"diffusive") == 0) {
> +    reflectionstyle = DIFFUSIVE;
> +    arginc = 6;
> +  } else if (strcmp(arg[3],"maxwell") == 0) {
> +    reflectionstyle = MAXWELL;
> +    arginc = 7;
> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
> +    reflectionstyle = CERCIGNANILAMPIS;
> +    arginc = 9;
> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
> +
> +  if (reflectionstyle != NONE) {
> +    seedfix = force->numeric(FLERR,arg[4]);
> +    if (seedfix > 900000000) error->all(FLERR,"Random seed must be less than 900000000");
> +    random = new RanMars(lmp,seedfix);
>
> This should be using |random = new RanMars(lmp,seedfix + comm->me);|. 
> Or else random numbers may be correlated between different MPI ranks. 
> See fix langevin.
>
> ------------------------------------------------------------------------
>
> In src/USER-MISC/fix_wall_stochastic.cpp 
> <https://github.com/lammps/lammps/pull/1691#discussion_r328323909>:
>
> > +
> +
> +  if (strcmp(arg[3],"diffusive") == 0) {
> +    reflectionstyle = DIFFUSIVE;
> +    arginc = 6;
> +  } else if (strcmp(arg[3],"maxwell") == 0) {
> +    reflectionstyle = MAXWELL;
> +    arginc = 7;
> +  } else if (strcmp(arg[3],"cercignanilampis") == 0) {
> +    reflectionstyle = CERCIGNANILAMPIS;
> +    arginc = 9;
> +  } else error->all(FLERR,"Illegal fix wall/stochastic command");
> +
> +  if (reflectionstyle != NONE) {
> +    seedfix = force->numeric(FLERR,arg[4]);
> +    if (seedfix > 900000000) error->all(FLERR,"Random seed must be less than 900000000");
>
> instead there should be a test not allowing a seed of 0 or negative. 
> See, e.g., fix langevin for comparison.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lammps/lammps/pull/1691?email_source=notifications&email_token=ANFYXKICREMAZZVVYP5WI53QLPCUJA5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCF6APTQ#pullrequestreview-293341134>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ANFYXKLNTAW3V7JZHNK6E3LQLPCUJANCNFSM4IZNCFFQ>.
>



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

sjplimp commented on doc/src/fix_wall_stochastic.txt:
```diff
@@ -0,0 +1,102 @@
+
+[Description:]
+
+This fix differs from the command fix wall/reflect which is equivalent
```
The first line(s) here should explain what this fix does.  Then explain how it is different from the related fix wall/reflect.

sjplimp commented on src/fix_wall_reflect.cpp:
```diff
@@ -38,106 +35,110 @@ FixWallReflect::FixWallReflect(LAMMPS *lmp, int narg, char **arg) :
-        wallstyle[nwall] = CONSTANT;
-        coord0[nwall] = force->numeric(FLERR,arg[iarg+1]);
-      }
+  if (strcmp(arg[2],"wall/reflect") == 0) { // child class can be stochastic
```
Can you invert this line so that the rest of the constructor does not need to 
be indented?  E.g. if the style is stochastic: return

toquydong replied:
> Can you invert this line so that the rest of the constructor does not need to
> be indented? E.g. if the style is stochastic: return

Your suggestion is indeed a simple coding way. My concern is that this stochastic model  is included in the optional package USER-MISC (except if you want to have it in the main directory) and in the future, there may be other wall models in other packages. In that case, the list can be long and is not optimal. The current implementation makes the wall/reflect independent from other wall models and those who do not use those walls, do not have to wonder what they are in the code. 

It is just a question of preference. If you think that including those lines is better, I will change the code.

akohlmey replied:
What @sjplimp suggested is indeed the preferred style.

sjplimp commented on src/fix_wall_reflect.h:
```diff
@@ -28,10 +28,13 @@ class FixWallReflect : public Fix {
   void init();
   void post_integrate();
-
+  enum{XLO=0,XHI=1,YLO=2,YHI=3,ZLO=4,ZHI=5};
```
@akohlmey I don't think we put enums in header files?

akohlmey replied:
We didn't in the past, but we should. I have already submitted multiple pull requests that implemented increments of this change. The header is the right place and helps to keep settings consistent. We had multiple bugs because people modified an enum in the implementation of the base class, but not in the derived classes. They incur no storage cost and are inside the name space (unlike defines).

sjplimp commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+    arginc = 9;
+  } else error->all(FLERR,"Illegal fix wall/stochastic command");
+
+  if (reflectionstyle != NONE) {
```
Again, you're indenting the entire rest of the method, which is needlessly complicated.
The reflectionstyle is never NONE.

toquydong replied:
Agreed.

sjplimp commented on src/USER-MISC/fix_wall_stochastic.cpp:
```diff
@@ -0,0 +1,256 @@
+  if (reflectionstyle != NONE) {
+    seedfix = force->inumeric(FLERR,arg[4]);
+    if (seedfix <=0) error->all(FLERR,"Random seed must be a postive number");
+    random = new RanMars(lmp,seedfix + comm->me);
```
This is never freed -> memory leak? You need a destructor method which does a delete random.

toquydong replied:
OK. I agree.

sjplimp commented:
Two other issues:
(1) I think the name of the new fix should probably be wall/reflect/stochastic.  Since
it derives from wall/reflect.  There are other places where the name wall/reflect is
searched for, e.g. in fix rigid to issue a warning if rigid bodies are used with wall/relect.
Since the same warning should be issued for the new fix stochastic, it will simply
searching for a match if it's wall/reflect/stochastic.
(2) This one applies to the original reflect and the new stochastic.  How should triclinic boxes
be handled?  I.e. the wall is not axis-aligned.  I don't think we currently do anything
to even check for triclinic and issue a warning or error.

akohlmey commented:
@toquydong this is now sitting for about 4 weeks with changes requested by @sjplimp and no reaction from you. Are you going to try to update your pull request? or is this one change request too many? ;-)

toquydong commented:
Dear Axel,
I am currently overloaded with work at my university. As I am just 
beginner, it takes me sometime to understand and figure out how to fix 
the issues raised by Steve (issue 1 for example). If someone can help me 
about this, I will appreciate and am willing to share the credit.

Quy Dong
Le 06/11/2019 18:03, Axel Kohlmeyer a écrit :
>
> @toquydong <https://github.com/toquydong> this is now sitting for 
> about 4 weeks with changes requested by @sjplimp 
> <https://github.com/sjplimp> and no reaction from you. Are you going 
> to try to update your pull request? or is this one change request too 
> many? ;-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lammps/lammps/pull/1691?email_source=notifications&email_token=ANFYXKKTQ7UOY472HXAOT4TQSL2HRA5CNFSM4IZNCFF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDHIIYI#issuecomment-550405217>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ANFYXKJJWQJITJZTGG4NHD3QSL2HRANCNFSM4IZNCFFQ>.
>



-- 
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

akohlmey commented:
> Dear Axel, I am currently overloaded with work at my university. As I am just beginner, it takes me sometime to understand and figure out how to fix the issues raised by Steve (issue 1 for example). If someone can help me about this, I will appreciate and am willing to share the credit. Quy Dong Le 

No problem about being busy, everybody is. And no problem with having problems to understand what was asked to change. Happens even to experienced people occasionally. So please ask @sjplimp for clarifications and explain what you understood he meant and where exactly you cannot follow. You can do this on the Github website, or - if this is too public for your taste - contact him directly at his sandia e-mail address. We don't worry about credit. All of us have put so much time into maintaining LAMMPS, it doesn't matter. We'll never get the amount of credit we have earned, anyway. ;-)

I was merely looking for a sign of life. It is always a good thing to put a message here just to acknowledge the comments that were made. ...and if you are busy, it is sufficient to just state that, and we'll keep quiet for a while.

sjplimp commented:
>Since you mention many things in the comment, it is not clear about what 
>I have to do. Does it simply mean changing the name of the source file 
>into fix_wall_reflect_stochastic.cpp instead of fix_wall_stochastic.cpp 
>etc... ?

@toquydong Thanks for working more on this.  My comments were aimed at minimizing
the changes to fix_wall_reflect.cpp, and putting more of your changes
in a new derived class fix wall_relfect_stochastic.cpp.  If you can do that,
it should be fine.

toquydong commented:
I have updated the code and the doc file.

sjplimp commented:
@toquydong I made some edits to the source files (parent wall/reflect and new child wall/reflect/stochastic).  Also modified the doc page for the new command.  Can you please check that everything looks OK and still works as you expect?  Also, if you can add your name and institution to the top of fix_wall_reflect_stochastic.cpp.  See the top of src/fix_nh.cpp for an examples.  Also add one line to src/USER-MISC/README with you name and contact info.  Then I think it is ready to release.  A bonus would be if you added an example dir like example/wall with some simple scripts that show the usage and difference between wall reflect and the new stochastic models.  I think this is a nice addition to LAMMPS.

sjplimp commented:
@akohlmey and @toquydong For now I left the new doc page for fix wall/reflect/stochastic as a txt file.  It could be converted to rst when @toquydong is happy with it.

sjplimp commented:
@akohlmey Please remind me, is the fix_wall_reflect.h file the right place for the enum settings used by both the parent and child classes?

akohlmey commented:
> @akohlmey Please remind me, is the fix_wall_reflect.h file the right place for the enum settings used by both the parent and child classes?

yes, and it should be defined *inside* the class, so that it remains in its namespace and does not collide with other enums.

akohlmey commented:
whitespace issues are removed, the fix added to `doc/src/fix.rst` and `doc/src/Commands_fix.rst`, spelling errors removed and false positives added. This now needs the confirmation from @sjplimp that his requested changes are all applied and it can be merged.

sjplimp commented:
I think that @toquydong should first verify the latest version works as advertised.

toquydong commented:
> @toquydong I made some edits to the source files (parent wall/reflect and new child wall/reflect/stochastic). Also modified the doc page for the new command. Can you please check that everything looks OK and still works as you expect? Also, if you can add your name and institution to the top of fix_wall_reflect_stochastic.cpp. See the top of src/fix_nh.cpp for an examples. Also add one line to src/USER-MISC/README with you name and contact info. Then I think it is ready to release. A bonus would be if you added an example dir like example/wall with some simple scripts that show the usage and difference between wall reflect and the new stochastic models. I think this is a nice addition to LAMMPS.

The edits are OK and working. The doc is OK too. I have added the information in the README file and cpp file. I also tried to give an example as you suggested. In the example, I use 10^6 time steps but in reality many more timesteps and average may be required to obtain reliable results.

toquydong commented:
> I think that @toquydong should first verify the latest version works as advertised.

I think it is OK.

sjplimp commented:
@toquydong Thanks for adding an example script.  Small size is fine.
